robot_smach_states/benchmark/benchmark_grasping.py

```diff
@@ -0,0 +1,224 @@
+    single.add_argument("support", type=str,
+                        help="ID of entity to grasp FROM ('grasp' entity is on-top-of this 'support' entity), eg. 'cabinet'")
+    single.add_argument("waypoint", type=str,
+                        help="ID of waypoint for start and end position of the robot")
+    single.add_argument("--inspect_from_area", type=str,
+                        help="What area of the support-entity to inspect from", default='in_front_of')
+
+    batch_config_fields = ['cls', 'support', 'waypoint', 'inspect_from_area', 'search_area']
+    batch = subparsers.add_parser(name='batch', description="Perform the single case repeatedly, "
+                                                            "taking configurations from a .csv file, "
+                                                            "with columns {}".format(','.join(batch_config_fields)))
+    batch.add_argument("--configuration", type=str, default='grasp_benchmark_config.csv')
+
+    args = parser.parse_args()
+    rospy.init_node("benchmark_grasping")
+
+    with open(args.output, 'a+') as csv_file:
+        reader = csv.DictReader(csv_file)
+
+        fields = ['timestamp', 'robot',
+                  'start_waypoint',
+                  'expected_class', 'observed_class',
+                  'id',
+                  'inspect_duration',
+                  'grab_duration',
+                  'x', 'y', 'z']
+        results_writer = csv.DictWriter(csv_file, fieldnames=fields, extrasaction='ignore')  # Ignore extra keys in the dicts
+
+        if reader.fieldnames != results_writer.fieldnames:
+            results_writer.writeheader()
+
+        robot = get_robot(args.robot)
+
+        if args.subcommand == 'single':
+            single_item(robot, results_writer,
+                        args.cls, args.support, args.waypoint, args.inspect_from_area,
+                        args.non_strict_class)
+        elif args.subcommand == 'batch':
+            with open(args.configuration) as config_file:
+                config_reader = csv.DictReader(config_file)
+                assert config_reader.fieldnames == batch_config_fields, "CSV file need fields {}".format(','.join(batch_config_fields))
+
+                configs = []
+
+                environment_config_description = ""
+                for config_row in config_reader:
+                    configs += [config_row]
+                    item_config_description = "Put a {cls} {search_area} the {support}"\
+                        .format(cls=config_row['cls'],
+                                support=config_row['support'],
+                                search_area=config_row['search_area'])
+                    rospy.logwarn(item_config_description)
+                    environment_config_description += item_config_description
+
+                records = []
+                for config_row in configs:
```

Comment:
Isn't this exactly the same as line 205 ("for config_row in config_reader")

Reply:
Yes. But I want to iterate over the items once first to notify the user and then iterate over them again to do the actual work